### DISCUSSION TITLE
sending OpenCV Mat from c++ to python, using buffered port

### DISCUSSION THREAD

**Discussion by valeriosperati** [robotology/community#306] [Q&A] [answered]
I'm having some problem sending an opencv Mat from  a c++ code to a python code, using a yarp buffered port. In some cases, depending on CV mat dimensions,  extra pixels seem to be added during the conversion from cv::Mat to yarp ImageOfPixel object, do not know why.

This is my c++ code.  
For communication I use a yarp::os::BufferedPort<Bottle> called portOUT;

```
// step 1: I create a foo mat 
cv::Mat foo;
foo.create(30, 40, CV_8UC3);
foo = cv::Scalar(255,0,255);

//step 2: I copy foo in yarp image
yarp::sig::ImageOf<sig::PixelRgb> img = yarp::sig::ImageOf<sig::PixelRgb>();
img.resize(foo.cols, foo.rows);
memcpy(img.getRawImage(), foo.data, sizeof(unsigned char) * foo.rows * foo.cols * foo.channels());

//step 3: I check the consistency between cv mat and yarp image;
// I would expect the same dimensions and no added padding 
cout<<img.width()<<img.height()<<img.getPadding()<<endl;
                            
//step 4: send image
Bottle& out = portOUT.prepare();
out.clear();
out.read(img);
portOUT.writeStrict();
```

Now, the code works if opencv mat has -for example- dimensions [30x40]  
I mean that in step 3 padding is 0 (as expected), and python side receives a bottle containing the correct image:
python side correctly reports the content of bottle as
```
[ 0 ] mat
[ 1 ] rgb
[ 2 ] 3 3600 8 40 30  //these are the correct dimensions: 40x30x3 = 3600
[ 3 ] ...the pixels of the image...
```
 
Nevertheless with other mat dimensions, for example [30x30] in step 1, I see that extra padding pixels are added during conversion in step 2: step 3 reports that padding is 6 (where do these extra pixels come from?). Python side receives a bottle with the following content:

[ 0 ] mat
[ 1 ] rgb
[ 2 ] 3 2880 8 30 30 // these dimensions are not correct: 30x30x3 = 2700, why do I get 2880?
[ 3 ] the pixels of the image...
 
I made some tests with other cv mat dimensions. It always fails when sent image is a square.
Thanks

**Comment by emilio-cartoni** [comment, marked as answer]
The following link code might be useful, as it shows how to convert from OpenCv to Yarp images:
http://www.yarp.it/opencv_2main_8cpp-example.html
However, wrapIplimage() method has a [big warning](http://www.yarp.it/classyarp_1_1sig_1_1Image.html#a1c9b867a3b4826966dc8baafc2ac11b3) when using BufferedPort, so maybe another approach is needed

> The wrapped image needs to exist for the rest of the lifetime of this oboject. Be careful if you use this method on objects read from or written to a BufferedPort, since the lifetime of such objects can be longer than you expect (see the documentation for yarp::os::BufferedPort::read, yarp::os::BufferedPort::prepare, and yarp::os::BufferedPort::write)

**Comment by Nicogene** [comment]
In the c++ code, why are you sending the image through a `yarp::os::Bottle` ?
I think it is better to do:
```c++
yarp::os::BufferedPort<yarp::sig::ImageOf<yarp::sig::PixelRgb>> outPort;
yarp::sig::ImageOf<yarp::sig::PixelRgb> & imgOut = outPort.prepare();

// change the image, copy data in it.

outPort.writeStrict();
```

**Comment by valeriosperati** [comment]
@Nicogene   I'm using a Bottle as sometimes I have to send an image, sometimes a different message (like a string).
@emilio-cartoni  the method wrapIplImage seems to work...with some mysterious issue: the extra padding pixels continue to be added but now the python side reconstructs a correct image;  

```
//creating a IplImage version of cv::Mat 
IplImage p = foo;
img.wrapIplImage(&p); 
```
img.getPadding() continues to report extra pixels.
